### PR TITLE
refactor: front-end type system

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
refactor type system. redo the type hierarchy - instead of distinction between "Primitives" and "Definitions", have single type system + ExprInfo

- simplify type hierarchy
- ~merge old and new type systems~ will do this in a follow up PR
- refactor builtins dir structure. merge `builtin_functions/` and `builtin_interfaces/` to `builtins/`.

### How I did it
rethinking of the approach; don't need two classes for each type, factor into VyperType + ExprInfo + VarInfo. handling of types in the regular namespace is generally by dispatching into special constructor (for structs and interfaces) or attributes (for enums) mode.

### Commit message
```
This commit refactors the front-end type system.

Currently, there are essentially two classes types for each vyper type.
"Primitives" represented raw types, and "Definitions" represent
the instantiation of types attached to each expression in a program.
"Definitions" additionally carry around annotation information related
to the expressions they tag, such as mutability and location info. This
system also handled dispatching into the correct routines for cases
where a raw type (e.g. "MyStruct") is used in expr land (vs an instance
of a type, e.g.  "my_struct"). This is important because certain types
are callable or attributable - specifically, constructors and enum
members (ex. `MyStruct()` and `MyEnum.FOO`).

This commit reworks the system by factoring the annotation info into
VyperType and ExprInfo. "Primitives" in the old system can be thought of
as VyperTypes, and "Definitions" in the old system are most similar to
the new ExprInfo. To handle cases where types can live in expr land,
this commit also clarifies that usage by promoting raw types into the
special, internal TYPE_T type (which can be thought of as "the type of
a type"). This commit also simplifies the inheritance structure of vyper
types.

Some other miscellaneous things in this commit:
- AST class hierarchy gets more refined - addition of ExprNode
- rename of `vyper/semantics/validation/` -> `vyper/semantics/analysis/`
- type system directory structure simplified
- merge of `builtin_functions/` and `builtin_interfaces/` to `builtins/`

Co-authored-by: z80 <z80dev@mlnl.finance>
```

### Cute Animal Picture

image

### CODE REVIEW

**Comment by charles-cooper** [comment]
this is close to ready -- passing tests and lint. still want to polish some things but mostly just small clarificatoins. in the meantime @fubuloubu do you want to take a first pass at reviewing?

**Review by skellet0r — APPROVED** [review]
Chad work ser.
I definitely like the reorganization done, adds clarity.

File: vyper/semantics/analysis/base.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+from vyper.exceptions import (
+    CompilerPanic,
+    ImmutableViolation,
+    StateAccessViolation,
+    VyperInternalException,
+)
+from vyper.semantics.types.base import VyperType
+
+
+class _StringEnum(enum.Enum):
```
Comment (fubuloubu) [inline vyper/semantics/analysis/base.py:15]: A lot of code just to avoid writing the same name twice?

**Review by fubuloubu — APPROVED** [review]

File: vyper/semantics/analysis/base.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/semantics/analysis/base.py:15]: that code was already there, just moved from vyper/semantics/types/utils.py: https://github.com/vyperlang/vyper/pull/2974/files#diff-2d8b16ee68553ace110d4070fab5d0f133beb1aef3a9acd1633951964600145bL41

**Pull request merged**